### PR TITLE
Upgrade addon-manager with kubectl apply

### PR DESCRIPTION
**Pull request by MrHohn**
The first step of kubernetes/kubernetes#33698.

Use `kubectl apply` to replace addon-manager's previous logic.

The most important issue this PR is targeting is the upgrade from 1.4 to 1.5. Procedure as below:
1. Precondition: After the master is upgraded, new addon-manager starts and all the old resources on nodes are running normally.
2. Annotate the old ReplicationController resources with kubectl.kubernetes.io/last-applied-configuration=""
3. Call `kubectl apply --prune=false` on addons folder to create new addons, including the new Deployments.
4. Wait for one minute for new addons to be spinned up.
5. Enter the periodical loop of `kubectl apply --prune=true`. The old RCs will be pruned at the first call.

Procedure of a normal startup:
1. Addon-manager starts and no addon resources are running.
2. Annotate nothing.
3. Call `kubectl apply --prune=false` to create all new addons.
4. No need to explain the remain.

Remained Issues:
- Need to add `--type` flag to `kubectl apply --prune`, mentioned [here](https://github.com/kubernetes/kubernetes/pull/33075#discussion_r80814070).
- This addon manager is not working properly with the current Deployment heapster, which runs [addon-resizer](https://github.com/kubernetes/contrib/tree/master/addon-resizer) in the same pod and changes resource limit configuration through the apiserver. `kubectl apply` fights with the addon-resizers. May be we should remove the initial resource limit field in the configuration file for this specific Deployment as we removed the replica count.

@mikedanese @thockin @bprashanth 

---

Below are some logical things that may need to be clarified, feel free to **OMIT** them as they are too verbose:
- For upgrade, the old RCs will not fight with the new Deployments during the overlap period even if they use the same label in template:
  - Deployment will not recognize the old pods because it need to match an additional "pod-template-hash" label.
  - ReplicationController will not manage the new pods (created by deployment) because the [`controllerRef`](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/controller-ref.md) feature.
- As we are moving all addons to Deployment, all old RCs would be removed. Attach empty annotation to RCs is only for letting `kubectl apply --prune` to recognize them, the content does not matter.
- We might need to also annotate other resource types if we plan to upgrade them in 1.5 release:
  - They don't need to be attached this fake annotation if they remain in the same name. `kubectl apply` can recognize them by name/type/namespace.
  - In the other case, attaching empty annotations to them will still work. As the plan is to use label selector for annotate, some innocence old resources may also be attached empty annotations, they work as below two cases:
    - Resources that need to be bumped up to a newer version (mainly due to some significant update --- change disallowed fields --- that could not be managed by the update feature of `kubectl apply`) are good to go with this fake annotation, as old resources will be deleted and new sources will be created. The content in annotation does not matter.
    - Resources that need to stay inside the management of `kubectl apply` is also good to go. As `kubectl apply` will [generate a 3-way merge patch](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/strategicpatch/patch.go#L1202-L1226).  This empty annotation is harmless enough.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34513)

### CODE REVIEW

**Comment by thockin** [comment]
Thanks @MrHohn for writing and @mikedanese for reviewing.  Sorry for the confusion last week.  I won't have time to look at this directly, but I am 100% confident you guys know my concerns on this and will handle it.

**Comment by mikedanese** [comment]
> +127 −529

Gotta love that diffstat. And the + is mostly comments :)

File: cluster/addons/addon-manager/kube-addons.sh [hunk 1]
```diff
@@ -35,35 +103,65 @@ function start_addon() {
...
 }
 
+# $1 resource type.
+function annotate_addons() {
+  local -r obj_type=$1;
+
+  # Annotate to objects already have this annotation should fail.
+  # Only try once for now.
+  # Keep it quite for not breaking the readability of logging.
+  ${KUBECTL} ${KUBECTL_OPTS} annotate ${obj_type} --namespace=${SYSTEM_NAMESPACE} -l kubernetes.io/cluster-service=true kubectl.kubernetes.io/last-applied-configuration='' --overwrite=false >/dev/null 2>&1
```
Comment (mikedanese) [inline cluster/addons/addon-manager/kube-addons.sh]: nit: Can you break this into a multiline command? It's quite long.

**Review by mikedanese — COMMENTED** [review]
This looks great. Are we going to want to convert most RC addons to deployments before v1.5?

**Comment by mikedanese** [comment]
> Call kubectl apply --prune=false to create all new addons.

I think we can even remove this step after v1.5 release.

File: cluster/addons/addon-manager/kube-addons.sh (hunk 1 above)
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons.sh]: Sure.

I'm a bit worried about keeping this quiet. If something go wrong, we will have more clues to look at if we let it shout. Maybe we should remove `>/dev/null 2>&1`?

Comment (mikedanese) [inline cluster/addons/addon-manager/kube-addons.sh]: I agree with your suggestion.

**Comment by MrHohn** [comment]
> I think we can even remove this step after v1.5 release.

Yeah. Do we need to maintain the old version to future upgrade (like 1.4->1.6)?

> Are we going to want to convert most RC addons to deployments before v1.5?

Do you mean whether converting most of the RCs or converting just kube-dns?

**Comment by mikedanese** [comment]
> Do you mean whether converting most of the RCs or converting just kube-dns?

I mean converting other applications as well. I don't think it makes sense to update RCs with kubectl apply after this goes in.

**Comment by MrHohn** [comment]
For upgrading the addon manager, we plan to remove some resource limit fields (for heapster and eventer containers) in [the configuration file for Heapster Deployment](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/cluster-monitoring/influxdb/heapster-controller.yaml) and let addon-resizers take care of the limit, otherwise the new addon manager will fight with the addon-resizer.

Would this be feasible? @piosz @gmarek

**Comment by piosz** [comment]
@MrHohn I'm fine with that assuming that Heapster will have resources set properly after creation. You should verify the logic of addon-resizer.

cc @Q-Lee

**Comment by thockin** [comment]
I don't know if all of the RC -> deployments can or should be done in 1.5,
and that's really the gate.

**Comment by gmarek** [comment]
I'm OK with having separate component that sets up limits for addons. PodNanny seems to be working just fine for now, but making it a central component will reduce total amount of memory used.

How do you plan to pass information about how much cpu/mem does a given add-on need per <something> (most likely Node) to addon-resizer?

I didn't look very closely at this PR, but I don't see how you deal with add-on deletions. Does kubectl apply  handle that for us?

**Comment by mikedanese** [comment]
> I don't know if all of the RC -> deployments can or should be done in 1.5,
> and that's really the gate.

Can you explan this? Do you think it's a bad to move addons to RCs?

**Comment by thockin** [comment]
My point was we have N>1 RCs ad addons now, yeah?  Are the people who own
those addons *all* ready to move to Deployments in 1.5?  If so, great,
ignore me.  The gate for getting rid of --prune=false is "all RCs are
converted to Deployment" not "passed v1.5".

**Comment by MrHohn** [comment]
@piosz Thanks for reply. I think current addon-resizer logic fits this plan.

@gmarek Sorry about the confusion. In my context I mean PodNanny = addon-resizer. We have no plan to make PodNanny a central component, so most logic of Heapster will remain the same.

Yes, the addons deletion is handled by `kubectl apply --prune`.

**Comment by MrHohn** [comment]
Thanks for the comment. There is no design doc yet. I will soon work on it if it is necessary. Though not pretty sure about the difference between design doc and README. 

For the things pointed out: (Will include this in README)
- The difficulty for changing an addon component might be somehow due to the ambiguous behavior of current Addon Manager --- only judging addons are changed or not based on their resource basenames and suffixes. After this upgrade any legit change in the addon configuration file will be reflected and the recommended way to update addons is through their configuration file on the master node. For Horizontal/Vertical Autoscale feature, the proper way to cooperate is leaving the related fileds empty in configuration file, so addon manager won't include it in the diff-patch.
- Sorry, the `kubernetes.io/cluster-service` label is still in used in both situations. I'm not quite aware why is this a problem?
- Saw similar issue before(#33894). This may or may not happen because this new Addon Manager only prune resource created / updated by `kubectl apply`. But I consider the behavior you mention is correct? All `cluster-service` should be under the Addon manager's control. If something need to be added as `cluster-service`, they should also be put in the same place on the master node?

**Comment by mikedanese** [comment]
This is labeled a WIP. What is left to do?

**Comment by MrHohn** [comment]
I also want to update the README and work on the Makefile, which currently using an old version release `kubectl`.

But I guess we can do that in other PRs?

**Comment by mikedanese** [comment]
Ya let's followup. It'll be good to let this soak.

/lgtm

**Comment by MrHohn** [comment]
Seems like the /lgtm is not working?

**Comment by mikedanese** [comment]
cc @kubernetes/sig-cluster-lifecycle

**Comment by jimmidyson** [comment]
Great to see this merged. Is there going to be an updated add-on manager image coming soon?

**Comment by MrHohn** [comment]
@jimmidyson The plan I know is to upgrade the addon manager at 1.5 release.

**Comment by jsafrane** [comment]
@MrHohn, can you update the image earlier? My PR #31617 needs kube-addon-manager to be able to install StorageClass instances, which is not possible in the old image as it has hardcoded list of object types it can handle: https://github.com/kubernetes/kubernetes/blob/v1.4.5/cluster/addons/addon-manager/kube-addon-update.sh#L480

**Comment by MrHohn** [comment]
@jsafrane Sorry for the inconvenience. I'm trying to update the Addon Manager earlier, but currently it is blocked on #35220, which adds a needed feature to `kubectl apply`.

**Pull request merged**